src/Api/Vault/AuthorizationHandlers/Collections/CollectionAuthorizationHandler.cs

```diff
@@ -83,20 +85,20 @@ public class CollectionAuthorizationHandler : BulkAuthorizationHandler<Collectio
     }
 
     private async Task CanCreateAsync(AuthorizationHandlerContext context, CollectionOperationRequirement requirement,
-        CurrentContextOrganization org)
+        CurrentContextOrganization? org)
     {
         // If false, all organization members are allowed to create collections
-        if (!org.LimitCollectionCreationDeletion)
+        if (org is { LimitCollectionCreationDeletion: false })
         {
             context.Succeed(requirement);
             return;
         }
 
         // Owners, Admins, Providers, and users with CreateNewCollections permission can always create collections
         if (
-            org.Type is OrganizationUserType.Owner or OrganizationUserType.Admin ||
-            org.Permissions is { CreateNewCollections: true } ||
-            await _currentContext.ProviderUserForOrgAsync(org.Id))
+            org is { Type: OrganizationUserType.Owner or OrganizationUserType.Admin } ||
+            org is { Permissions.CreateNewCollections: true } ||
+            await _currentContext.ProviderUserForOrgAsync(_targetOrganizationId))
```

Comment:
If the user is a providerUser, this sproc is now called twice: once in the `HandleRequirementAsync` method and once here.

We could just move it exclusively to the main method, but I like each switch case being self-contained. And it's kind of awkward, e.g.

```cs
if (org is null && isProvider)
{
  context.Succeed(...);
  return;
}
else if (org is null)
{
  return;
}

// switch statement here which exclusively deals with orgUsers
```

I think this shows that the structure of the handler isn't quite as clear as it could be.

What if we removed the providerUser check in the public method, and then each private method was laid out like this:

```cs
private async Task CanCreateAsync(...)
{
  if (org is not null)
  {
  // check organizationUser requirements
  // early return if success
  }

  if (providerUser)
  {
  // success
  }

// if it's not authorised above then it just drops through the if statements and is unauthorized
}
```

At the moment the org logic is too long to nest it like this, but @r-tome extracted the `manageableCollections` logic in https://github.com/bitwarden/server/pull/3360 which makes it more concise. I think with both your refactors it could work quite well.

Reply:
I do like your suggestion and went ahead and refactored the private methods to something similar and removed the provider check in the public method. It should also work nicely with @r-tome's `HasCollectionAccess()` refactor that you mentioned.